### ISSUE TITLE
[epic] add more options to `service create` / `service update`

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#25303]
The `service create` and `service update` commands do not support all options that `docker run` / `docker create` supports. Some options are not implemented yet, whereas other options may either *not* be implemented (because they don't make sense in the context of a *service*, or are not portable / cross platform).

We should add more options for services, *however* instead of blindly copying every option, we should make sure the options are implemented properly, which may require using different names for the options and/or different kind of values.

I tried to create an overview of all options on `docker run`, and to match them with the `docker service create` options we currently have; I may have missed some, or made the wrong "translation", so input is welcome here

Status                | Issue      | `docker run`                          | `docker service`                                                                       |  Notes
----------------------|------------|---------------------------------------|----------------------------------------------------------------------------------------|---------------------------------
                      |            |       `--add-host`                    |                                                                                        | 
                      |            |   `-a, --attach`                      |                                                                                        | 
                      |            |       `--blkio-weight`                |                                                                                        | 
                      |            |       `--blkio-weight-device`         |                                                                                        | 
                      |            |       `--cap-add`                     |                                                                                        | 
                      |            |       `--cap-drop`                    |                                                                                        | 
                      |            |       `--cgroup-parent`               |                                                                                        | 
                      |            |       `--cidfile`                     |                                                                                        | 
 ?                    |            |       `--cpu-percent`                 | `--limit-cpu`??                                                                        | 
 ?                    |            |       `--cpu-period`                  | `--limit-cpu`??                                                                        | 
 ?                    |            |       `--cpu-quota`                   | `--limit-cpu`??                                                                        | 
 ?                    |            |   `-c, --cpu-shares`                  | `--limit-cpu`??                                                                        | 
 ?                    |            |       `--cpuset-cpus`                 | `--limit-cpu`??                                                                        | 
 ?                    |            |       `--cpuset-mems`                 | `--limit-cpu`??                                                                        | 
 :white_check_mark:   |            |   `-d, --detach`                      |                                                                                        |  `-d` is the default
                      |            |       `--detach-keys`                 |                                                                                        | 
                      |            |       `--device`                      |                                                                                        | 
                      |            |       `--device-read-bps`             |                                                                                        | 
                      |            |       `--device-read-iops`            |                                                                                        | 
                      |            |       `--device-write-bps`            |                                                                                        | 
                      |            |       `--device-write-iops`           |                                                                                        | 
                      |            |       `--disable-content-trust`       |                                                                                        | 
                      |            |       `--dns`                         |                                                                                        | 
                      |            |       `--dns-opt`                     |                                                                                        | 
                      |            |       `--dns-search`                  |                                                                                        | 
                      |            |       `--entrypoint`                  |                                                                                        | 
 :white_check_mark:   |            |   `-e, --env`                         | `-e, --env`                                                                            | 
                      | #24712     |       `--env-file`                    |                                                                                        | 
                      |            |       `--expose`                      |                                                                                        | 
                      |            |       `--group-add`                   |                                                                                        | 
                      |            |       `--health-cmd`                  |                                                                                        | 
                      |            |       `--health-interval duration`    |                                                                                        | 
                      |            |       `--health-retries`              |                                                                                        | 
                      |            |       `--health-timeout duration`     |                                                                                        | 
                      | #24877     |   `-h, --hostname`                    |                                                                                        | 
                      |            |   `-i, --interactive`                 |                                                                                        | 
                      |            |       `--io-maxbandwidth`             |                                                                                        | 
                      |            |       `--io-maxiops uint`             |                                                                                        | 
                      |            |       `--ip`                          |                                                                                        | 
                      |            |       `--ip6`                         |                                                                                        | 
                      |            |       `--ipc`                         |                                                                                        | 
                      |            |       `--isolation`                   |                                                                                        | 
                      |            |       `--kernel-memory`               |                                                                                        | 
 :white_check_mark:   |            |   `-l, --label`                       | `--container-label`                                                                    | 
                      |            |       `--label-file`                  |                                                                                        | 
                      |            |       `--link`                        |                                                                                        | 
                      |            |       `--link-local-ip`               |                                                                                        | 
 :white_check_mark:   |            |       `--log-driver`                  | `--log-driver`                                                                         | 
 :white_check_mark:   |            |       `--log-opt`                     | `--log-opt`                                                                            | 
 :white_check_mark:   |            |       `--mac-address`                 |                                                                                        | 
                      |            |   `-m, --memory`                      | `--limit-memory`                                                                       | 
                      |            |       `--memory-reservation`          |                                                                                        | 
                      |            |       `--memory-swap`                 |                                                                                        | 
                      |            |       `--memory-swappiness`           |                                                                                        | 
 :white_square_button:|            |       `--name`                        |                                                                                        |  NOTE: `--name` sets the *service* name, not the container's name
 :white_square_button:|            |       `--network`                     | `--network`                                                                            |  Does not support `host` networking
                      | #24787     |       `--network-alias`               |                                                                                        | 
                      |            |       `--no-healthcheck`              |                                                                                        | 
                      |            |       `--oom-kill-disable`            |                                                                                        | 
                      |            |       `--oom-score-adj`               |                                                                                        | 
                      |            |       `--pid`                         |                                                                                        | 
                      |            |       `--pids-limit`                  |                                                                                        | 
                      | #24862     |       `--privileged`                  |                                                                                        | 
 :white_check_mark:   |            |   `-p, --publish`                     | `-p, --publish`                                                                        | 
                      |            |   `-P, --publish-all`                 |                                                                                        | 
                      |            |       `--read-only`                   |                                                                                        | 
 :white_check_mark:   |            |       `--restart`                     | `--restart-condition`, `--restart-delay`, `--restart-max-attempts`, `--restart-window` | 
                      |            |       `--rm`                          |                                                                                        | 
                      |            |       `--runtime`                     |                                                                                        | 
                      | #25209     |       `--security-opt`                |                                                                                        | 
                      |            |       `--shm-size`                    |                                                                                        | 
                      |            |       `--sig-proxy`                   |                                                                                        | 
                      |            |       `--stop-signal`                 |                                                                                        | 
                      |            |       `--storage-opt`                 |                                                                                        | 
                      | #25209     |       `--sysctl`                      |                                                                                        | 
                      |            |       `--tmpfs`                       |                                                                                        | 
                      |            |   `-t, --tty`                         |                                                                                        | 
                      | #25209     |       `--ulimit`                      |                                                                                        | 
 :white_square_button:|            |   `-u, --user`                        | `-u, --user`                                                                           |  Does not support group / gid
                      |            |       `--userns`                      |                                                                                        | 
                      |            |       `--uts`                         |                                                                                        | 
 :white_check_mark:   |            |   `-v, --volume`                      | `--mount`                                                                              |  UX improvement needed (add `-v` flag?)
 :white_check_mark:   |            |       `--volume-driver`               | `--mount`                                                                              |  UX improvement needed (add `-v` flag?)
                      |            |       `--volumes-from`                |                                                                                        | 
 :white_check_mark:   |            |   `-w, --workdir`                     | `-w, --workdir`                                                                        |

**Comment by thaJeztah** [comment]
ping @stevvooe @aluzzardi @icecrime @tiborvass

**Comment by yongtang** [comment]
@thaJeztah I saw on #25276 and the above list that `--net=host` is not supported. That could be a very useful option in some deployment scenarios. I am wondering if there are any technical challenges that prevent the usage of `--net=host`?

**Comment by yongtang** [comment]
@thaJeztah For `--mount` there is an open issue related: #25253

**Comment by thaJeztah** [comment]
@yongtang I know that for some options there's additional discussion needed; e.g. `--network=host` can have some side effects (how does it work in combination with load-balancing? also `--network=host` can be considered insecure)

**Comment by yongtang** [comment]
Thanks @thaJeztah for the explanation.

**Comment by justincormack** [comment]
@thaJeztah `--ip` and `--ipv6` don't make sense either as there are multiple containers.

**Comment by thaJeztah** [comment]
Good call, I'll edit

**Comment by stevvooe** [comment]
- `-i` is not necessary.
- `--groud-add` will be `--group`.
- `--link` and `--link-local-ip` are never going to be implemented for services.
- `--publish-all` doesn't make sense.
- `--cidfile` doesn't make sense
- `--volumes-from` is never going to happen

**Comment by thaJeztah** [comment]
@stevvooe agreed on most; in doubt on `--publish-all`, although I can see it possibly being difficult to implement (ports aren't known until the image was pulled / inspected, perhaps that's more difficult in Swarm mode networking)

**Comment by thaJeztah** [comment]
Thinking about `--link` - providing that may allow for container/service scoped aliasing; but we should analyze use-cases and see if it makes sense

**Comment by stevvooe** [comment]
> @stevvooe agreed on most; in doubt on --publish-all, although I can see it possibly being difficult to implement (ports aren't known until the image was pulled / inspected, perhaps that's more difficult in Swarm mode networking)

This just doesn't make sense. For the most part, ports are already exported on the container's IP. In swarm services, this would bind all ports across the cluster, which isn't useful. The other issue is that the manager has _no idea_ what these ports actually are, without inspecting the image. Exported cluster ports should be considered individually and carefully.

> Thinking about --link - providing that may allow for container/service scoped aliasing; but we should analyze use-cases and see if it makes sense

This is provided by overlay networking and network aliasing. In general, `--link` should be considered deprecated. Interdependent containers don't yield reliable architectures. Dependencies should be resolved before creation (based on a rendevouz value or service name/alias).

**Comment by thaJeztah** [comment]
Yes, implementing `--network-alias` will probably solve most use cases; we need to think of a UX for that, perhaps that may require a syntax similar to `--mount` to provide per-network options (see https://github.com/docker/docker/issues/24787#issuecomment-233853394)

**Comment by hvpareja** [comment]
I've also seen the options `--env-add` and `--env-rm` in `docker service update`. They replace `--env` flag. They seems to be missing in your table.

**Comment by thaJeztah** [comment]
@hvpareja hm, yes, `service update` has different flags to add/remove options for flags that can be provided multiple times, not sure if we need to include that here, but thanks for the suggestion

**Comment by kyhavlov** [comment]
+1 for being able to set `net=host` for services, that would be very useful for high performance things that don't want the docker network overhead

**Comment by tzz** [comment]
In our environment we'd like to use `--add-host`, is that a possibility? Thank you!

**Comment by stevvooe** [comment]
@tzz How do you use `--add-host` in your environment? Do you use it to route to external services or internal services?

The reason I ask is we are curious if this belongs on the network, as part of a service or requires a separate abstraction to manager to external service discovery. @mrjana

**Comment by tzz** [comment]
@stevvooe internal services with external visibility. Our problem is that `service.wanted` resolves in DNS to an external IP address, which is appropriate for some connections, but other connections need the internal IP address for the same service to avoid NAT and routing. Other possibilities for us are dnsmasq and keeping both addresses in DNS, but either way that's a lot more work than `--add-host` and less dynamic at service creation time. I hope that's helpful.

**Comment by stevvooe** [comment]
@tzz Thanks for the explanation!

Would it be correct to say that you want this at the cluster, network or service-level? For example, will have different values for `service.wanted` depending if you are in different services?

**Comment by tzz** [comment]
@stevvooe for stable environments (production) the lookup can be stable at the cluster or network level. That's our most important use case. For unstable (dev and QA) environments we may need to change the lookup at the service level more dynamically, but that's a less important use case for us.

**Comment by drajen** [comment]
I'm pulling my hair that docker service won't allow -t. Certain apps can't live without a TTY.

**Comment by stevvooe** [comment]
@drajen Could you file a bug on SwarmKit?

Out of curiosity, what applications are you running as services that require a TTY?

**Comment by drajen** [comment]
@stevvooe I was running Samba in foreground mode which apparently requires a tty. Check the samba-publicshare image on docker hub. You may easily write a wrapper around it but there might be more breakage out there.

**Comment by stevvooe** [comment]
@drajen That is a bummer. I'm really sorry we missed that. I wish someone spoke up in the RC period.

I've submitted https://github.com/docker/swarmkit/pull/1370 and https://github.com/docker/engine-api/pull/364 for service support. We'll have to follow that up with the analogous flags in `docker service`.

**Comment by drajen** [comment]
@stevvooe thanks for giving this attention!

**Comment by vikstrous** [comment]
I'm so confused "command" for service doesn't override "entrypoint". What does `[COMMAND] is the "entrypoint" of the container` mean? If I have a container I want to run as a service that already has an entrypoint set in the image, I don't see a way to override it.

**Comment by sm0svx** [comment]
+1 on network=host. The lack of this option is basically a showstopper for me since my application connect out to a specific hardware unit using TCP. The unit then send data back using UDP to the originating IP address for the TCP connection. This works well in docker run with net=host but of course fail using the docker0 bridge in swarm mode.

Is there another standard way to solve this without using network=host?  New network bridge directly bridged to the physical network interface and using valid external IP-addresses for the containers maybe?

**Comment by stevvooe** [comment]
@vikstrous I'm not sure where you read this, but the statement "`[COMMAND] is the "entrypoint" of the container`" very confusing and technically wrong. Where did you read this? Let's file a bug to add some clarification there.

This is a long time confusion over a poor choice of wording in the context of Unix history. First off, we have the concept of a _command_, which is simply an array of strings passed to an executable. This is how we execute a process. All of this language is in the context of assembling a command.

For the purposes of docker, `Entrypoint` is the default command for an image. It also has a field `Cmd`, which are the default arguments to `Entrypoint`. So, without any arguments, the _command_ is `Entrypoint` + `Cmd` from the image. The standard invocation of a docker container or service is simply to pass a `Cmd` to the engine and that will be used as the full _command_ or be appended to the `Entrypoint` in the image, if it exit to assemble the full command.

`COMMAND` for a service should be `ARGS`. The reason it is `[COMMAND]` in the help output, is that it may be the first entry of the _command_ if `Entrypoint` is no defined for the image. This cannot be parsed at service creation, since we don't know anything about the image, so we don't know if it can be overridden.

It sounds like your question is asking about why `--entrypoint` is not provided. The reasoning behind that is it makes less sense in the context of services to run arbitrary commands inside of a container. Typically, `--entrypoint` is used for interactive debugging, which is the domain of running individual containers, not services. If there is a solid case for a use case to override `Entrypoint`, we may allow it, as well.

**Comment by vikstrous** [comment]
@stevvooe My quote is from this the description of this issue. It's in the table. I wasn't sure what the story around the omission of `--entrypoint` was. This makes sense now. I can work around this limitation for our services for now, but my opinion is that `--entrypoint` should be added to the CLI because it exists in the API (with a different name, but still with the ability to override the entrypoint).

**Comment by fanux** [comment]
We need `--net=host` ,  the overlay model will drop our network speed and consume more CPU.  We have two network interface, the overlay model will bing a lots of problems to us.  If swarm not support the `--net=host` ,  we can't use it at all....

**Comment by thaJeztah** [comment]
@fanux can you open a separate issue, and ping me there? Then I'll link the issue from here. `--net=host` may need some discussion, and probably better to do that on a separate issue. If possible,  please include some more details on your use case

**Comment by alexellis** [comment]
Hi, there @thaJeztah please can you update the grid with issue https://github.com/docker/docker/issues/25885 for --add-cap ?

**Comment by dhiltgen** [comment]
> It sounds like your question is asking about why --entrypoint is not provided. The reasoning behind that is it makes less sense in the context of services to run arbitrary commands inside of a container. Typically, --entrypoint is used for interactive debugging, which is the domain of running individual containers, not services. If there is a solid case for a use case to override Entrypoint, we may allow it, as well.

I can think of a number of images that have multiple tools within them that are all related, and the "default" one is wired up as the entrypoint for ease-of-use.  One example is https://hub.docker.com/_/docker/  Perhaps this isn't generally the best practice, but I'm +1 with @vikstrous that lacking the ability to override on service definition forces people to re-generate images that would otherwise be usable as services.

**Comment by thaJeztah** [comment]
@vikstrous @stevvooe sorry, the faulty description is my mistake; I think at some point in the design discussions, we talked about if both `ENTRYPOINT` and `CMD` are present in an image, that basically `ENTRYPOINT` is `command`, and `CMD` is `args`. Following that, `docker service create foo -bar` would lead to `foo` being the *entrypoint*  and `-bar` being an *arg*, but it's not the case

**Comment by dhiltgen** [comment]
> @vikstrous @stevvooe sorry, the faulty description is my mistake; I think at some point in the design discussions, we talked about if both ENTRYPOINT and CMD are present in an image, that basically ENTRYPOINT is command, and CMD is args. Following that, docker service create foo -bar would lead to foo being the entrypoint and -bar being an arg, but it's not the case

Actually, I take back what I said... it looks like maybe this is just a doc confusion...  The entrypoint of the image is actually overwritten based on the CMD fed to the service definition.

**Comment by fanux** [comment]
Hi @thaJeztah #25873 is a new issue for --net=host.

**Comment by cmingxu** [comment]
IMHO,  --net=host options should be available only for service with global mode

**Comment by cmingxu** [comment]
--add-host, --network-alias, --link need more clarification on how are they functioned and how to use them.

**Comment by cmingxu** [comment]
--env-file on which side the file stay,  manager? node? or where the command was issued?

**Comment by stascode** [comment]
How about --network=bridge/my_custom_bridge network. Consider a scenario where you need run a bunch of containers in the global mode being able to communicate with each other within the worker host. This option would be somewhat similar to the experimental bundles and stacks.

**Comment by gittycat** [comment]
@dhiltgen I'm not seeing that. With `docker service create myimage COMMAND`, `COMMAND` doesn't override the `ENTRYPOINT`, which is an issue with some images as you've noted.

@stevvooe Thanks for your many comments in multiple issues about this. This whole ENTRYPOINT/CMD/COMMAND is so inanely confusing. I was really hoping that specifying a command at service creation would overwrite ENTRYPOINT. That doesn't seem to be the case unfortunately.

**Comment by stevvooe** [comment]
@stascode @cmingxu Adding those isn't as simple as adding the flags. I'd recommend taking that discussion to another issue. They need to be thought through in detail.

@dhiltgen Many images have tools embedded, but do they make sense to run as services?

@gittycat I hope my comments have helped. All the `COMMAND` fields (from help output) are simply passed as args to `ENTRYPOINT`. This is the way `docker run` has always worked. There is a flag `--entrypoint` on `docker run` that allows one to override it. If you build an image without an entrypoint, `COMMAND` will be used as the effective entrypoint.

**Comment by gittycat** [comment]
@stevvooe I understand that. The problem as you know is that there isn't a way to override entrypoint when creating a service. The only work around is create our own custom images to reset `ENTRYPOINT:[]` when it's already defined in the base image.

**Comment by stevvooe** [comment]
@gittycat Ok, great. We are on the same page.

It sounds like you want `--entrypoint` similar to what is in `docker run`.

**Comment by gittycat** [comment]
My preferences are (in that order)

1. Make the service command overwrite the image `ENTRYPOINT` and `CMD`.
    Eg: `docker service create someimage echo "hello"` should always echo "hello" independent of what's in the image ENTRYPROINT and CMD. 

2. Add `--entrypoint` to the `docker service create|update` api. That's still confusing but at least we can overwrite the entrypoint, which isn't the case now.

**Comment by hervenicol** [comment]
Hey all,

I'd rather like second choice from @gittycat , as it's the way it has always been with either docker standalone or classic swarm.

I, as well as lots of people, have already created images that take advantage of this entrypoint / command logic. If the «command» arguments to «service create» override the entrypoint, that's going to be a mess !

**Comment by gittycat** [comment]
@hervenicol As @stevvooe mentioned in some other post, this is a chance to get things right by removing the confusing mess that is entrypoint/cmd. Once `entrypoint` is added as an option to `docker service`, we'll be stuck with it forever. 

I strongly prefer the first option; override entrypoint/cmd by the service command.

**Comment by hervenicol** [comment]
If docker chooses to go that way, then entrypoint should be also deprecated on standalone docker.


But I think the entrypoint / command is convenient, once you understand it.

For instance, if you look at how to run a consul image (https://github.com/gliderlabs/docker-consul/tree/legacy):
`docker run progrium/consul -server -bootstrap`
You don't have to care about what's the path/name of the app, all you want is to know its parameters.

And if you're looking for some help, you run
`docker run progrium/consul --help`
or even
`docker run progrium/consul`

I think it is VERY convenient, and this behaviour would be broken if entrypoint/command dissociation was removed.
I'd rather go fore some disambiguation in the documentation.

**Comment by gittycat** [comment]
@hervinicol Yes, you describe the whole reason for having an `entrypoint`. It makes it possible to treat a container as if it were an executable. It's nice at first but I personally have stopped using this. I much prefer having things explicitly visible in docker-compose or in command line scripts. For each image that I use, I always end up having to go to the Dockerfile to see what the ENTRYPOINT/CMD are.

**Comment by stevvooe** [comment]
@gittycat @hervenicol While we've pretty much deprecated the concept in the Service API, deprecating entrypoint in `docker run` and `docker service create` is pretty much a non-starter.

Likely, the solution will revolve around @gittycat's option 2.

**Comment by shenshouer** [comment]
any idea for swarm mode support host networking ?

**Comment by Tchinmai7** [comment]
A --security-opt option in service will make Checkpoint Restore easier!. so far, Trying CR with docker service always fails during the checkpoint time!

**Comment by gittycat** [comment]
The `--security-opt` is also needed by Elasticsearch. Elasticsearch errors on startup `seccomp unavailable: your kernel is buggy and you should upgrade`. The [workaround given](https://github.com/docker-library/elasticsearch/issues/98#issuecomment-230316529) is to start containers using `--security-opt=seccomp=unconfined` but that's not available for services. What is a suggested workaround if any?

**Comment by thaJeztah** [comment]
@gittycat https://github.com/docker/docker/issues/25209 is for tracking that option for services, perhaps you can add your use case there as well. One challenge will be to add a custom profile for services, as that would mean that the actual file containing the profile is present on each node (unless we store its content in the service definition @justincormack?)

**Comment by pugnascotia** [comment]
This might be a different issue, but I'd like to see an option on `service create` that specifies a delay before a container is consider in-service. The use case is where you're doing a rolling update to a service, and it needs some amount of warm up time before it can handle requests. If you're relying on the routing mesh for load balancing and you get an error because a container isn't ready yet, you can't tell if the service is broken or a particular upstream container isn't available because you're using a virtual IP. Hope that makes sense.

**Comment by hervenicol** [comment]
@pugnascotia , today I guess you can do it with update-delay.

But for a better solution, IMO that should be handled through healthchecks. If your container says it's healthy only when it's «warm» and ready to serve requests, it should be the time when:
- this actual container is enabled in the IPVS
- next container starts being deployed (maybe after waiting for update-delay).

**Comment by pugnascotia** [comment]
update-delay doesn't help in this case because that (AFAICT) controls the time gap between finishing the update of one container and starting the update of the next. I agree that some kind of healthcheck mechanism is the way to go.

**Comment by stevvooe** [comment]
@pugnascotia @hervenicol This issue is about discussing the flags from `docker run` in the context of `docker service create`. For generic feature requests, I'd recommend opening up an issue. There are a number of issues regarding IPVS, so checking out some those might help you find what you're looking for.

**Comment by richardjq** [comment]
@thaJeztah please add --time to this list. As in 'docker stop --time=120 foo' or in swarm-speak 'docker service rm --time=120 foo'

I'm having to run apps unclustered because 10 seconds isn't long enough for a clean shutdown.

Thanks!

**Comment by dnephin** [comment]
@richardjq this option already exists as `--stop-grace-period`

**Comment by richardjq** [comment]
@dnephin Awesome! I didn't think to look on the create/update command. Thanks!

**Comment by g0nx** [comment]
Whats happen with --hostname option? I think it is very important because there are a lot of scenarios in which your containers need to have hostname like example.com because the certificates. For example I need to communicate a rest service with a oauth service and it is only possible throw hostname oauth.example.com (I have a crt with *.example.com within this container).

I  hope that is clear...

**Comment by thaJeztah** [comment]
@g0nx see https://github.com/docker/docker/issues/24877 for the discussion on that one (it's linked in the table at the top)

**Comment by g0nx** [comment]
I have already read this thread but I can't see if this option going to implement to the next version or not... And the most important thing, when?

**Comment by thaJeztah** [comment]
@g0nx that decision will be made on that issue, not here (so far, no decision was made)

**Comment by g0nx** [comment]
@thaJeztah  ok! Thank you very much for your attention!

**Comment by cezarsa** [comment]
Regarding healthchecks, would simply exposing `--health-*` and `--no-healthcheck` flags in `docker service create` be an acceptable solution? I imagine encoding them inside the `swarm.ContainerSpec` instance and just using them when creating new containers for the service. I could work on a PR for this if no one has started yet.

**Comment by errordeveloper** [comment]
@thaJeztah was nice meeting you in Berlin last week! Could you please add https://github.com/docker/swarmkit/issues/1605 for `--pid=host`... Thanks 👍

**Comment by thaJeztah** [comment]
@errordeveloper likewise! I added it to the table

**Comment by merijnv** [comment]
Somewhere this was asked:

> Do you use it to route to external services or internal services?

We want to have an MQ, which is running on a native OS rather than within docker, and reach that by name. We do not control the DNS resolving, so I would like to add 'my-queuing-server' as an ip-address in /etc/hosts as seen by the container.

**Comment by BretFisher** [comment]
@thaJeztah should we also add lack of `affinity` in services to this list?

RE: https://github.com/docker/swarmkit/issues/1442

**Comment by stevvooe** [comment]
Update on host-level networking: we are discussing a simplified model in https://github.com/docker/swarmkit/pull/1645. It would allow containers to be exposed on both the host-level network and any swarm overlay networks that may apply. Still need to figure out what `--network=host` might mean, but it could be as simple as "the network is off and the ports are exposed". Port locations would be exposed through swarm managers.

**Comment by mostolog** [comment]
Hi.

Does docker services support multiples ports exposed/published for a service?
Seems docker service create --publish 8080 --publish 8888 doesn't work...
Regards

**Comment by stevvooe** [comment]
@mostolog Could you please file a separate issue? We need to keep the discussion here focused.

**Comment by kleptog** [comment]
One thing I don't see on this list in `docker service update --network-add` and `--network-rm`. These are being tracked at https://github.com/docker/swarmkit/issues/1029.

**Comment by adityacs** [comment]
Any update on how to provide cpu-quota and cpu-period values while creating a service in swarm mode?

**Comment by ventz** [comment]
@thaJeztah In regards to the:

"--ip and --ip6 don't make sense either as there are multiple containers."

It makes sense to have them set the VIP address. There are many cases where the IP needs to be statically set/available ahead of time/specially provisioned/etc.

**Comment by thaJeztah** [comment]
@ventz can you open a new issue for that, and describe the use case(s) there? I'll link the issue in the table then. Out of curiosity; those cases cannot use a _hostname_ only an _IP address_? (include that in the use-case description as well)

**Comment by mostolog** [comment]
Does _name templating_ (https://github.com/docker/docker/blob/master/docs/reference/commandline/service_create.md#create-services-using-templates) work within compose files for docker deploy? Should it? Is documented somewhere?

Is there any way that documents what is "id, slot and all that fields"?

**Comment by stevvooe** [comment]
@mostolog It should but that is off topic for this particular issue.

**Comment by piersharding** [comment]
Hi -

I am working on an HPC project that would need to achieve CPU pinning for containers so that specific tasks run on the appropriate CPU as well as speak directly to the host network adapter (a topic covered well by others above).  This could be achieved by having a head node task (perhaps on the swarm manager node) that uses docker swarm to run an agent container on each compute node launched as a service and then each agent container could then launch the appropriately configured task containers locally using docker run --cpuset-cpus="1,2" etc. However, it would be much cleaner/nicer to manage the entire deployment process through docker swarm so that orchestration features would not need to be implemented in the agent nodes simplifying the entire orchestration control structure.

On this basis I would like to vote for the inclusion of the --cpuset-cpus parameter.

Thanks,
Piers Harding.

**Comment by thaJeztah** [comment]
@piersharding could you open a separate issue, then i'll link to that issue from the table above

**Comment by piersharding** [comment]
@thaJeztah thanks - I have created issue #30477 .

Cheers,
Piers Harding.

**Comment by mrueg** [comment]
Is the table missing --init? Would be interesting to have that in services as well.

**Comment by thaJeztah** [comment]
@mrueg good catch; it was added in 1.13, so not a thing when this table was created originally

**Comment by genki** [comment]
I think --volumes-from could be applicable to the volumes of services that are mode=global.

**Comment by ventz** [comment]
@thaJeztah Referencing: https://github.com/docker/docker/issues/30963

This is the perfect use case for --ip for the service:

Mariadb Galera (mysql DB cluster) -- needs to exchange the nodes involved ahead of time. You can use name, but the --hostname makes it so that the node itself resolves to its second IP vs what the rest of the nodes resolve it to, which is first ip. If we can specify the IP statically, they can be pre-configured ahead of time by IP.

**Comment by PatrickLang** [comment]
@thaJeztah can you add this issue for --isolation to your master list ? https://github.com/docker/docker/issues/31616

**Comment by thaJeztah** [comment]
@PatrickLang added 👍

**Comment by seongkki** [comment]
Any news on #31595?

**Comment by pascalandy** [comment]
I can’t stop thinking about the [Zero-downtime deployments with rolling upgrades](https://github.com/moby/moby/issues/30321) topic. For me Moby would be basically perfect with this feature is working (+ a stable stateful solution. I count on the Infinit team for this.)

What’s the game plan to empower users to do this?

Is a flag like `rolling-upgrade` will be set to `true` by default?

Cheers!
Pascal

**Comment by daixiang0** [comment]
Could add  `--device-cgroup-rule` option for service?

**Comment by carlfischer1** [comment]
@thaJeztah housekeeping: ```--isolation``` is implemented

**Comment by mbentley** [comment]
I haven't see much action on this issue but another thing that I have not seen solved yet from classic swarm is the ability to schedule containers with affinities.  Not sure if that is out of scope of this epic or not but that is a bit of a gap in being able to have one process per container in Swarm mode.

**Comment by malkir** [comment]
Please consider adding a --sysctl flag to Docker Service. It being missing nullifies the use of Swarm in high performance environments when having to, for example, wait 60s for a TIME_WAIT connection to clean out. Tens of thousands of TIME_WAIT connections can exist within an individual container and it severely affects performance. A modification of net.ipv4.tcp_fin_timeout or net.ipv4.tcp_tw_reuse on the host network namespace is not adequate when it makes no affect to the containers themselves.

**Comment by sirlatrom** [comment]
I'm looking forward to the day I can no longer create a distributed fork bomb using `docker service create bash ':(){ :|:& };:; wait'`.

**Comment by thaJeztah** [comment]
@jefflill could you open a separate issue for that, then I'll link it from the table at the top

**Comment by jefflill** [comment]
@thaJeztah: Done.  Thanks.

**Comment by hholst80** [comment]
`-i` is still missing from the list. it was said to be included from https://github.com/moby/moby/issues/32300

We need this to work efficiently with discrete resources like GPUs.

**Comment by thaJeztah** [comment]
> What extra information is needed for the implementation of --interactive? Why cannot the implementation used by docker-compose be used here (stdin_keep: true)?

When running a swarm service, the container itself can run on any node in the swarm cluster; the API won't be able to connect to a container that's on any other host than the manager that is connected to.

**Comment by deepio** [comment]
devices! 
Something like this would be nice. 

```yaml
version: "3.9?"
services:
  gpu-celery:
     ...
    deploy:
      placement:
        constraints:
          - node.labels.queue == GPU
      devices:
        - "/dev/nvidia0:/dev/nvidia0 "
        - "/dev/nvidiactl:/dev/nvidiactl "
        - "/dev/nvidia-uvm:/dev/nvidia-uvm"
```

**Comment by ahmadfarisfs** [comment]
I think adding --env-file options to docker service update should be fine. Right now it's only on docker service create

**Comment by thaJeztah** [comment]
@ahmadfarisfs better to discuss on https://github.com/moby/moby/issues/31595, which also lists some of the reasons it's more problematic for `docker service update`.

**Comment by kaysond** [comment]
@thaJeztah - are there any updates on `--privileged` support? The PRs (https://github.com/moby/swarmkit/pull/1722 & #32801) seem to be dead. What I really care about is setting namsepaced sysctl's via a swarm service (see #43769) but I couldn't find any other way of doing so.

**Comment by markperri** [comment]
Has there been any updates on --storage-opt to set the max size of the container?
